### ISSUE TITLE
boost::optional example doesn't work on older Boost

### ISSUE THREAD

**Issue by bmerry** [pybind/pybind11#847]
#### Issue description

The [documentation](http://pybind11.readthedocs.io/en/master/advanced/cast/stl.html#c-17-library-containers) describes how to wrap `boost::python`. It fails to compile on Boost 1.58 due to the line `value = {};` which was only fixed to work in [Boost 1.63](http://www.boost.org/doc/libs/1_64_0/libs/optional/doc/html/boost_optional/relnotes.html).

#### Reproducible example code

C++:
```c++
#include <iostream>
#include <pybind11/pybind11.h>
#include <pybind11/stl.h>
#include <pybind11/numpy.h>

namespace py = pybind11;

#define USE_BOOST
#ifdef USE_BOOST
#include <boost/optional.hpp>
using boost::optional;

namespace pybind11 { namespace detail {
    template <typename T>
    struct type_caster<boost::optional<T>> : optional_caster<boost::optional<T>> {};
}}
#else
#include <experimental/optional>
using std::experimental::optional;
#endif

static void func(optional<py::array_t<double>> x)
{
    if (x)
        std::cout << "Called with array of size " << x->size() << '\n';
    else
        std::cout << "Called with None\n";
}

PYBIND11_PLUGIN(use_opt)
{
    py::module m("use_opt", "Test wrapping of optional");
    m.def("func", &func);
    return m.ptr();
}
```

Python:
```python
#!/usr/bin/env python
import numpy as np
import use_opt

use_opt.func(np.array([1.0, 2.0]))
use_opt.func(None)
```

There has been some discussion in gitter about approaches to fix this. I'm going to assign to myself to create a PR.

**Comment by bmerry** [comment]
Looks like I can't actually assign to myself, presumably because I'm not a committer. I'll work on a PR next week though.

**Comment by jagerman** [comment]
Some potentials from the gitter discussion:
- a `type_traits`-type class to yield the `std::nullopt`-equivalent class
- a SFINAE `reset()` method with `= {};` fallback (or perhaps vice versa)
- Adding a `typename NullT` template parameter (e.g. set to ` = std::nullopt_t` for `std::optional`).  This would break backwards compatibility, however, if someone out there has already used `optional_caster` with Boost 1.63+.

**Comment by bmerry** [comment]
I've added example code to reproduce the issue. It seems that Boost 1.58 sometimes works with `value = {}`, depending on the wrapped class.

**Comment by dean0x7d** [comment]
Fixed via #874.
